src/ripple/app/sidechain/Federator.cpp

```diff
@@ -815,6 +819,13 @@ Federator::init(
     ticketRunner_.setRpcChannel(false, sidechainListener_);
     mainDoorKeeper_.setRpcChannel(mainchainListener_);
     sideDoorKeeper_.setRpcChannel(sidechainListener_);
+
+    heartbeatTimer =
+        std::make_unique<boost::asio::deadline_timer>(ios, heartbeatInterval);
+    heartbeatTimer->async_wait(
+        [self = shared_from_this()](boost::system::error_code const& ec) {
```

Comment:
`this` should work here, right, since `heartbeatTimer` will be destructed before its containing object?

Reply:
If the heartbeat timer gets destroyed is the callback also destroyed? I'm used to using `shared_from_this` in asio because the lifetime of the executor can outlive the lifetime of the captured object (especially during shutdown). You might be right that a timer destroys the call back, but we'd have to check. Let's leave it as-is for now.

Reply:
> if the heartbeat timer gets destroyed is the callback also destroyed?

yes I believe that is the case, see [this](https://www.boost.org/doc/libs/1_75_0/doc/html/boost_asio/reference/basic_deadline_timer/_basic_deadline_timer.html)

Reply:
That cancels the timer, which means the callback is still exists and may be called. If the callback is running when the `heartbeatTimer` is deleted then `this` would be become dangling. Even if it isn't in the middle on a call, it will still call the callback with `ec` set to `operation_aborted` and a dangling `this`. In the current implementation (and most sane implementations) we check `ec` before accessing `self`. Still, I'd rather not have the pointer dangle at all. Yeah, I think we need the `shared_from_this` here.

Reply:
>  which means the callback is still exists and may be called. 

I don't think so. It says: This function destroys the timer, **cancelling any outstanding asynchronous wait operations associated with the timer as if by calling cancel**.
I'm pretty sure the cancel will happen in the deadline_timer destructor, which means the callback will be called then, when the class owning the deadline_timer is still present. I don't see how `this` can be dangling.
Just discussing to make sure we both agree on what would happen, I have no issue with using `shared_from_this`.

Reply:
The captured variable `dangling` points to a dangling variable in the code below. The destructor will cause the operation to be cancelled, but that doesn't mean the callback is called before the destructor is finished running.

Even if cancellation isn't an issue, destroying the object in the middle of a callback would still be an issue.

```
#include <boost/asio.hpp>
#include <boost/asio/deadline_timer.hpp>
#include <iostream>
#include <limits>
#include <memory>
#include <vector>

int
main ()
{
    boost::asio::io_context ios;
    {
        boost::posix_time::seconds const heartbeatInterval{5};
        boost::asio::deadline_timer t (ios, heartbeatInterval);
        int dangling = 0;
        t.async_wait ([&dangling] (boost::system::error_code const& ec) {
            std::cout << "ec: " << ec << '\n';
            std::cout << "dangling: " << dangling << '\n';
        });
    }
    ios.run ();

    return 0;
}
```

Reply:
```
 #include <boost/asio.hpp>
#include <boost/asio/deadline_timer.hpp>
#include <iostream>
#include <limits>
#include <memory>
#include <vector>

struct DestructorProbe
{
    int x{2};
    boost::asio::deadline_timer t;

    DestructorProbe(boost::asio::io_context &ios) :
        t(ios, boost::posix_time::seconds{5})
    {
        t.async_wait ([this] (boost::system::error_code const& ec) {
            std::cout << "ec: " << ec << '\n';
            std::cout << "dangling: " << x << '\n';
        });
    }
        
    ~DestructorProbe ()
    {
        x = 42;
    }
};

int
main ()
{
    boost::asio::io_context ios;
    {
        DestructorProbe dangling(ios);
    }
    ios.run ();

    return 0;
}
```

Reply:
Just leaving a note: we talked in person; the `shared_pointer` is needed. resolving this conversation.